### PR TITLE
fix(vue-renderer): use `modulepreload` for modern mode spa generate

### LINKED ISSUE #5486: Modern build still loads "nomodule" scripts
### Version

[2.6.1](https://github.com/nuxt.js/releases/tag/2.6.1)

### Reproduction link

[http://example.com/sorry_it_is_impossible](http://example.com/sorry_it_is_impossible)

### Steps to reproduce

Have a simple nuxt.js SPA.

`nuxt.config.js`

```js
module.exports = {
    mode: "spa",

    modern: "client",

...
```

Build it:
```sh
NODE_ENV=production ./node_modules/.bin/nuxt generate
```

### What is expected ?

See the picture. Only the green scripts are expected to be loaded.

### What is actually happening?

The "nomodule" scripts are still loaded.

### Additional comments?

How would I completely drop support of the non-modern browsers?

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c9013">#c9013</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by clarkdo**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Fix #5486

In generate, use `modulepreload` modern resources instead of `preload` legacy resources for avoiding downloading both `modern` and `legacy` resources.

@pi0 I'm using `_generate` in context, not sure if it's a stable flag, please review 😄  

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/vue-renderer/src/spa-meta.js [hunk 1]
```diff
@@ -30,8 +30,9 @@ export default class SPAMetaRenderer {
     return vm.$meta().inject()
   }
 
-  async render({ url = '/', req = {} }) {
-    const cacheKey = `${req.modernMode ? 'modern:' : 'legacy:'}${url}`
+  async render({ url = '/', req = {}, _generate }) {
+    const modern = req.modernMode || _generate
```
Comment (pi0) [inline packages/vue-renderer/src/spa-meta.js]: How `nuxt generate` implies modern mode is enabled?

Comment (clarkdo) [reply packages/vue-renderer/src/spa-meta.js]: `req.modernMode` is set through middleware which is not called in `generate`.

With this fix, the behaviour is same as `vue-cli` doing

`modulepreload`: all modern js, only in modern browsers
`<script type='module'>`: all modern js, only in modern browsers
`<script nomodule>`: all legacy js, only in legacy browsers

Comment (Atinux) [reply packages/vue-renderer/src/spa-meta.js]: I agree on this, we want to make sure he built with the `modern` mode too no?

Comment (clarkdo) [reply packages/vue-renderer/src/spa-meta.js]: Oh, sorry, I missed the `modern` is not enabled case

Comment (clarkdo) [reply packages/vue-renderer/src/spa-meta.js]: @pi0 @Atinux I added `options.modern` checking, only generate `modern` resources in modern mode.

File: packages/vue-renderer/src/spa-meta.js [hunk 2]
```diff
@@ -30,8 +30,9 @@ export default class SPAMetaRenderer {
     return vm.$meta().inject()
   }
 
-  async render({ url = '/', req = {} }) {
-    const cacheKey = `${req.modernMode ? 'modern:' : 'legacy:'}${url}`
+  async render({ url = '/', req = {}, _generate }) {
```
Comment (pi0) [inline packages/vue-renderer/src/spa-meta.js]: This is extracting `renderRoute` `context`. Are you sure this context also includes `_generate`? I think it should be accessed with `this.options._generate`

Comment (clarkdo) [reply packages/vue-renderer/src/spa-meta.js]: I found this flag from https://github.com/nuxt/nuxt.js/blob/dev/packages/generator/src/generator.js#L204

The reason why I did't use `this.options._generate` is: `generate` in `build` or programmatic `generate` don't have `_generate`

Comment (pi0) [reply packages/vue-renderer/src/spa-meta.js]: Nice find 😊 I think that's an older flag but you are right

**Review by pi0 — APPROVED** [review]

**Review by pi0 — APPROVED** [review]

**Comment by Atinux** [comment]
One last question, will it work with `nuxt build --spa --modern`?

**Comment by clarkdo** [comment]
> One last question, will it work with `nuxt build --spa --modern`?

Yes, it works

**Pull request merged**